### ISSUE TITLE
Broken NTLM proxy authentication

### ISSUE THREAD

**Issue by vlet** [haproxy/haproxy#1140]
## Detailed description of the problem

Hi.

We use haproxy as load-balancer in front of several Cisco WSA proxy servers. WSA requires proxy authentication, supported Kerberos and NTLM schemes, but only kerberos proxy auth works fine via haproxy, NTLM authentication is broken. We have checked several version: 2.3.5, 2.2.8, 2.0.20 - same results. Only working version is haproxy 1.8.x (1.8.28).

## Expected behavior

Working NTLM proxy auth

## Steps to reproduce the behavior

Example of curl trace 

```
$ curl -v --cacert WSA01.crt --proxy-ntlm --proxy-user user:pass --proxy http://haproxy.srv:8080 -o /dev/null --no-progress-meter  https://ya.ru
*   Trying 10.0.0.221:8080...
* Connected to haproxy.srv (10.0.0.221) port 8080 (#0)
* allocate connect buffer!
* Establish HTTP proxy tunnel to ya.ru:443
* Proxy auth using NTLM with user 'user'
> CONNECT ya.ru:443 HTTP/1.1
> Host: ya.ru:443
> Proxy-Authorization: NTLM TlRMT....AAA=
> User-Agent: curl/7.74.0
> Proxy-Connection: Keep-Alive
> 
< HTTP/1.1 407 Proxy Authentication Required
< mime-version: 1.0
< date: Fri, 19 Feb 2021 14:39:48 MSK
< via: 1.1 WSA01.SRV:8080 (Cisco-WSA/11.8.2-009)
< content-type: text/html
< proxy-authenticate: NTLM TlRMTVNTU.........AAAA=
< proxy-connection: keep-alive
< content-length: 0
< 
* Establish HTTP proxy tunnel to ya.ru:443
* Proxy auth using NTLM with user 'user'
> CONNECT ya.ru:443 HTTP/1.1
> Host: ya.ru:443
> Proxy-Authorization: NTLM TlRMTVNTU.....................MwMTA=
> User-Agent: curl/7.74.0
> Proxy-Connection: Keep-Alive
> 
* Proxy CONNECT aborted
* CONNECT phase completed!
* Closing connection 0
curl: (56) Proxy CONNECT aborted
```

## Do you have any idea what may have caused this?

Haproxy close connection to WSA proxy backend in the middle of negotiation despite of explicit 'proxy-connection: keep-alive' header issued by wsa.

## What is your configuration?

```
global
		log /dev/log    local0
		log /dev/log    local1 notice
		chroot /var/lib/haproxy
		stats socket /run/haproxy/admin.sock mode 660 level admin expose-fd listeners
		stats timeout 30s
		user haproxy
		group haproxy
		daemon
		maxconn 400000

defaults
		timeout connect 5000
		timeout client  50000
		timeout server  50000

frontend Cisco-WSA-Proxy
		bind 0.0.0.0:8080
		mode http
		option http-keep-alive
		maxconn 400000
		# Internet Explorer and NTLM with HTTPS sites
		option http-use-proxy-header
		option accept-invalid-http-request
		default_backend Cisco-WSA-Servers

backend Cisco-WSA-Servers
		mode http
		option http-keep-alive
		balance source
		hash-type consistent
		option redispatch
		option forwardfor
		option prefer-last-server

		# Backends
		server WSA01 10.0.0.1:8080
                #server WSA02 10.0.0.2:8080
```

## Output of `haproxy -vv` and `uname -a`

```
HA-Proxy version 2.3.5-1~bpo10+1 2021/02/06 - https://haproxy.org/
Status: stable branch - will stop receiving fixes around Q1 2022.
Known bugs: http://www.haproxy.org/bugs/bugs-2.3.5.html
Running on: Linux 4.19.0-14-amd64 #1 SMP Debian 4.19.171-2 (2021-01-30) x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -O2 -fdebug-prefix-map=/build/haproxy-2.3.5=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1 USE_SYSTEMD=1
  DEBUG   = 

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -CLOSEFROM +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=2).
Built with OpenSSL version : OpenSSL 1.1.1d  10 Sep 2019
Running on OpenSSL version : OpenSSL 1.1.1d  10 Sep 2019
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.3
Built with network namespace support.
Built with the Prometheus exporter as a service
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.32 2018-09-10
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 8.3.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2
            fcgi : mode=HTTP       side=BE        mux=FCGI
       <default> : mode=HTTP       side=FE|BE     mux=H1
       <default> : mode=TCP        side=FE|BE     mux=PASS

Available services : prometheus-exporter
Available filters :
	[SPOE] spoe
	[CACHE] cache
	[FCGI] fcgi-app
	[COMP] compression
	[TRACE] trace

```

**Comment by capflam** [comment]
Confirmed. I'm able to reproduce the bug, I'm on a fix. I works as expected on the 2.4. Thanks

**Comment by chipitsine** [comment]
is there simple way to launch NTLM auth for automated tests ?

**Comment by capflam** [comment]
I don't know but in fact, the bug is about the CONNECT management in the H1 mux. If something else than a 200-ok is returned in response to a CONNECT request, the connection is closed. So it is no related to the NTLM.

**Comment by chipitsine** [comment]
NTLM itself is fragile. I would like to have some e2e automated test

**Comment by capflam** [comment]
It should be fixed in 2.3. The commit is ecae6e9 :
```
commit ecae6e9e3c3313ed7567b0b5799dbabfc4168da6 (HEAD -> master, haproxy.org/master)
Author: Christopher Faulet <cfaulet@haproxy.com>
Date:   Mon Feb 22 08:11:59 2021 +0100

    BUG/MEDIUM: mux-h1: Fix handling of responses to CONNECT other than 200-ok
    
    For a CONNECT request, if the tunnel establishment is refused by the server,
    the connection is always closed on the client side. This happen because we
    fail to detect the end of the tunnel. Now, when a reponse other than 200-ok
    is received, the request is switch back to MSG_DONE state and the end of the
    transaction is handled as a classical request/response exchange.
    
    This patch should fix the issue #1140. It must be backported as far as
    2.0. There is no upstream commit ID because tunnel management was already
    fixed in a non-backportable way in 2.4.
```
